Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/fragments/PermissionsAndCameraSettingFragment.kt

```diff
@@ -16,70 +16,96 @@
-                                    item.cameraId, item.size.width, item.size.height, item.fps))
-                }
+class PermissionsAndCameraSettingFragment : Fragment() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        val cameraManager =
+                requireContext().getSystemService(Context.CAMERA_SERVICE) as CameraManager
+
+        val cameraList = enumerateHighSpeedCameras(cameraManager)
+
+        lowestSetting = findLowestCameraSetting(cameraList)
+
+        if (hasPermissions(requireContext())) {
+            // If permissions have already been granted, proceed
+            Navigation.findNavController(requireActivity(), R.id.fragment_container).navigate(
+                    PermissionsFragmentDirections.actionPermissionsFragmentToCameraFragment2(
+                    lowestSetting.cameraId, lowestSetting.size.width, lowestSetting.size.height, lowestSetting.fps))
+        } else {
+            // Request camera-related permissions
+            requestPermissions(PERMISSIONS_REQUIRED, PERMISSIONS_REQUEST_CODE)
+        }
+    }
+
+    /** Selects the lowest camera fps and the lowest image capture resolution available among them.*/
+    private fun findLowestCameraSetting(cameraList: List<CameraInfo>): CameraInfo {
+        var bestLowestSettingSeen = cameraList.get(0)
+        for (setting in cameraList) {
+            if (setting.fps <= bestLowestSettingSeen.fps &&
+                    setting.size.width <= bestLowestSettingSeen.size.width &&
+                    setting.size.height <= bestLowestSettingSeen.size.height) {
+                bestLowestSettingSeen = setting
             }
         }
+        return bestLowestSettingSeen
+    }
 
+    override fun onRequestPermissionsResult(
+            requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == PERMISSIONS_REQUEST_CODE) {
+            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // Takes the user to the success fragment when permission is granted
+                Navigation.findNavController(requireActivity(), R.id.fragment_container).navigate(
+                        PermissionsFragmentDirections.actionPermissionsFragmentToCameraFragment2(lowestSetting.cameraId, lowestSetting.size.width, lowestSetting.size.height, lowestSetting.fps))
+            } else {
+                Toast.makeText(context, "Permission request denied", Toast.LENGTH_LONG).show()
+            }
+        }
     }
 
     companion object {
 
+        /** Convenience method used to check if all permissions required by this app are granted */
+        fun hasPermissions(context: Context) = PERMISSIONS_REQUIRED.all {
+            ContextCompat.checkSelfPermission(context, it) == PackageManager.PERMISSION_GRANTED
+        }
+
+        private lateinit var lowestSetting: CameraInfo
```

Comment:
Done.
> Please abstract a new method for detecting the lowest setting and navigating to the other activity immediately.

This can't be done at the moment because in this fragment, the UI will be waiting for the permission. If it already has that, it moves with the paramaters of the `lowestSetting`

Currenlty this fragment is intented for asking permissions only and I migrated code from selector fragment. I'm planning to migrate this to AnalyserFragment(new) when merging of CameraActivity and AnalyserActivity happens. 

Until then I'm logically moving code around while combining selector and permissions fragment.

Reply:
I think maybe you misunderstood...

Make a new function:

```
private fun startCamera() {
       val cameraManager =
                requireContext().getSystemService(Context.CAMERA_SERVICE) as CameraManager
        val cameraList = enumerateHighSpeedCameras(cameraManager)
        val lowestSetting = findLowestCameraSetting(cameraList)

        Navigation.findNavController(requireActivity(), R.id.fragment_container).navigate(
                    PermissionsFragmentDirections.actionPermissionsToSelector())
                    PermissionsFragmentDirections.actionPermissionsFragmentToCameraFragment(
                            lowestSetting.cameraId, lowestSetting.size.width, lowestSetting.size.height, lowestSetting.fps))
}
```

And call it in the two places you currently have similar code.

That way there is no need to store lowestSetting.

Reply:
Ah, I see.
Fixed in the commit.